File: lib/world_cup_tracker/matrix/brightness.ex
defmodule WorldCupTracker.Matrix.Brightness do
  @moduledoc """
  Time-of-day brightness scaling + gamma correction, applied by `Matrix.Rgb565`
  when encoding the device frame. The canonical framebuffer stays full-range;
  correction is applied only at encode time. The display's local timezone drives
  auto-dimming (full by day, dim overnight).
  """
  @db Tz.TimeZoneDatabase

  @day 1.0
  @night 0.25

  # Precompute gamma 2.2 as a 0..255 -> 0..255 lookup tuple at compile time.
  @gamma (for v <- 0..255 do
            round(:math.pow(v / 255, 2.2) * 255)
          end)
         |> List.to_tuple()

  @doc "Brightness scalar 0.25..1.0 for a local hour-of-day (float 0.0..24.0)."
  def for_hour(h) when h >= 8.0 and h <= 18.0, do: @day
  def for_hour(h) when h >= 6.0 and h < 8.0, do: lerp(@night, @day, (h - 6.0) / 2.0)
  def for_hour(h) when h > 18.0 and h <= 22.0, do: lerp(@day, @night, (h - 18.0) / 4.0)
  def for_hour(_h), do: @night

  @doc "Gamma-2.2 corrected value for a 0..255 channel."
  def gamma(v) when v in 0..255, do: elem(@gamma, v)

  @doc "Gamma-correct each channel of an {r,g,b}, then scale by `scalar` (0..1), clamped 0..255."
  def apply({r, g, b}, scalar) do
    {scale(r, scalar), scale(g, scalar), scale(b, scalar)}
  end

  @doc "Brightness scalar for a display timezone at epoch-ms `now`."
  def for_zone(tz, now_ms) do
    dt = now_ms |> DateTime.from_unix!(:millisecond) |> DateTime.shift_zone!(tz, @db)
    for_hour(dt.hour + dt.minute / 60.0)
  end

  defp scale(c, scalar), do: min(255, round(gamma(c) * scalar))
  defp lerp(a, b, t), do: a + (b - a) * t
end

File: src/brightness.cpp
#include "brightness.h"
#include <math.h>
#include <time.h>

namespace {
uint8_t gammaLut[256];

inline float lerp(float a, float b, float t) { return a + (b - a) * t; }

const float DAY = 1.0f;
const float NIGHT = 0.25f;
}  // namespace

namespace Brightness {

void init() {
  for (int v = 0; v < 256; v++) {
    gammaLut[v] = (uint8_t)lroundf(powf(v / 255.0f, 2.2f) * 255.0f);
  }
}

// Mirrors Brightness.for_hour/1 exactly.
float forHour(float h) {
  if (h >= 8.0f && h <= 18.0f) return DAY;
  if (h >= 6.0f && h < 8.0f) return lerp(NIGHT, DAY, (h - 6.0f) / 2.0f);
  if (h > 18.0f && h <= 22.0f) return lerp(DAY, NIGHT, (h - 18.0f) / 4.0f);
  return NIGHT;
}

float forNow(int64_t nowMs) {
  time_t secs = (time_t)(nowMs / 1000);
  struct tm lt;
  localtime_r(&secs, &lt);
  return forHour(lt.tm_hour + lt.tm_min / 60.0f);
}

Rgb correct(Rgb c, float scalar) {
  auto scale = [&](uint8_t v) -> uint8_t {
    long s = lroundf(gammaLut[v] * scalar);
    return s > 255 ? 255 : (uint8_t)s;
  };
  return {scale(c.r), scale(c.g), scale(c.b)};
}

}  // namespace Brightness
